exercises/palindrome-products/src/example.c

```diff
@@ -1,36 +1,104 @@
 #include "palindrome_products.h"
-#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <limits.h>
 
-/* Return true if the product of a and b is a palindrome, false otherwise */
-bool is_palindrome_product(int a, int b)
+int palindrome(int n);
+void addfactors(factor_t ** p, int i, int k);
+void check_alloc(void *p);
+void swap(int *i, int *k);
+
+product_t *get_palindrome_product(int from, int to)
 {
-   int ret = 0, prod = a * b;
-   while (prod) {
-      ret *= 10;
-      ret += prod % 10;
-      prod /= 10;
+   product_t *res = (product_t *) malloc(sizeof(product_t));
+   check_alloc(res);
+
+   res->error[MAXERR - 1] = '\0';
+   res->smallest = INT_MAX;
+   res->largest = INT_MIN;
+   res->factors_lg = NULL;
+   res->factors_sm = NULL;
+
+   if (from > to) {
+      snprintf(res->error, MAXERR - 1,
+               "invalid input: min is %i and max is %i", from, to);
+      return res;
    }
-   return (a * b) == ret;
-}
 
-product_t get_palindrome_product(int min, int max)
-{
-   product_t ret;
-   ret.smallest = 0;
-   ret.largest = 0;
-   bool largest = false;
-
-   for (int i = min; i <= max; i++) {
-      for (int j = min; j <= max; j++) {
-         if (is_palindrome_product(i, j) == true) {
-            if (largest == false) {
-               ret.smallest = i * j;
-               largest = true;
-            } else {
-               ret.largest = i * j;
+   int i, k, n;
+   for (i = from; i <= to; i++)
+      for (k = i; k <= to; k++)
+         if (palindrome(n = i * k)) {
+            if (n <= res->smallest) {
+               res->smallest = n;
+               addfactors(&res->factors_sm, i, k);
+            } else if (n >= res->largest) {
+               res->largest = n;
+               addfactors(&res->factors_lg, i, k);
             }
          }
-      }
+
+   if ((res->smallest == INT_MAX) || (res->largest == INT_MIN)) {
+      snprintf(res->error, MAXERR - 1,
+               "no palindrome with factors in the range %i to %i", from, to);
+      return res;
+   }
+   return res;
+}
+
+void free_ll(struct factors *p)
+{
+   if (p == NULL)
+      return;
+   if (p->next != NULL)
+      free_ll(p->next);
+   free(p);
+}
+
+void free_product(product_t * p)
+{
+   free_ll(p->factors_lg);
+   free_ll(p->factors_sm);
+   free(p);
+}
+
+void addfactors(factor_t ** p, int i, int k)
+{
+   int n = i * k;
+   if ((*p == NULL) || (((*p)->factor_a) * (*p)->factor_b != n)) {
+      free_ll(*p);
+      *p = NULL;
+   }
+
+   factor_t *tmp = (factor_t *) malloc(sizeof(factor_t));
+   check_alloc(tmp);
+
+   tmp->next = *p;
+   tmp->factor_a = i;
+   tmp->factor_b = k;
+   *p = tmp;
+}
+
+int palindrome(int n)
+{
+   if (n < 0)
+      n *= -1;
+   int r = 0;
+   int nn = n;
+
+   /* inverse n in r */
+   while (n >= 1) {
+      r = (r * 10) + (n % 10);
+      n /= 10;
+   }
+
+   return (nn == r);
+}
+
+void check_alloc(void *p)
+{
+   if (p == NULL) {
+      fprintf(stderr, "Memory error!\n");
+      exit(1);
```

Comment:
Factoring out the allocation check probably saves some typing but I would tend to shy away from using `exit()`... what do others think on this point?

A pattern that i often used to keep `malloc` checking code to a  vertical minimum is to do something like:

```c
product_t *res = NULL;

if ( (res = malloc(sizeof(product_t)) == NULL )
{
    // print out some message ...
    return 1;
}
```

Reply:
Most of my work is done without a heap, but when there's an unrecoverable error of this sort, I tend to think that getting out as quickly as possible is often a good choice. So I'm not opposed to using `exit()`, but I think that either solution works in this environment.